### PR TITLE
dev/CH-1375 consolidated tesl clients

### PR DESCRIPTION
**Pull request by aghuna**
Consolidated Tesl Client and added a StubTeslClient that will allow users of Tesl to test code that sends messages.

### CODE REVIEW

File: http_client/client.go [hunk 1]
```diff
@@ -1,37 +1,318 @@
...
+// Error fields might be modified by this method.
+func (response *ResponsePayload) extractResponseFromResult(values []reflect.Value) {
+	// Do nothing if we have only one return value, this is a plain message and there is
+	// no response to be returned
+	if len(values) == 1 {
+		return
+	}
+
+	// First extract the response
+	data, err := json.Marshal(values[0].Interface())
```
Comment (svmehta) [inline http_client/client.go]: do we need to check the length of values is not 0 here - is the assumption that the return type is (interface{}, error) or error enforceable?

File: http_client/client.go [hunk 2]
```diff
@@ -1,37 +1,318 @@
...
+// reflect values sent are in either (interface{}, error) format or (error).  The Error
+// might be modified by this method
+func (response *ResponsePayload) extractErrorFromResult(values []reflect.Value) {
+	// Assume that the error is the last field in the payload
+	idx := 0
+	if len(values) == 2 {
+		idx = 1
+	}
+
+	errResult := values[idx].Interface()
```
Comment (svmehta) [inline http_client/client.go]: same as comment above

File: http_client/client.go [hunk 3]
```diff
@@ -1,37 +1,318 @@
...
+		} else {
+			// If this somehow happens, it must've bypassing previous validation
+			// and is probably a bug
+			err = deeperror.New(3121059081, "Invalid basename "+key, nil)
+		}
+		if err != nil {
+			return deeperror.New(860762838, "Cannot register handler "+key, err)
+		}
+	}
+	err := client.ConsumerWrapper.Start()
```
Comment (svmehta) [inline http_client/client.go]: not a right now concern, but i was wondering whether the logic of setting up the wait group could be moved to the client itself...

see my note here: https://github.com/collectivehealth/beacon/blob/master/main.go#L134

**Comment by svmehta** [comment]
Looks good overall - thanks for the detailed comments. If you can update the readme as well (https://github.com/collectivehealth/tesl/tree/master/http_client) that'd be great. I feel like a quick example there is useful even with godocs since spinning up the documentation server locally in order to read high level setup might be a pain for new devs (and old devs!). We should start work on hosting all of the docs as well.

File: http_client/client.go (hunk 1 above)
Comment (aghuna) [inline http_client/client.go]: Enforced here on registering handler https://github.com/collectivehealth/tesl/pull/28/files#diff-0deb4851715a4fb854bdf379f91bc977R301

File: http_client/client.go (hunk 2 above)
Comment (aghuna) [inline http_client/client.go]: Also enforced on registering – assume that the handlers conform after they are registered.

File: http_client/client.go (hunk 1 above)
Comment (svmehta) [inline http_client/client.go]: :+1:

File: http_client/client.go (hunk 3 above)
Comment (aghuna) [inline http_client/client.go]: Agreed – more complicated though.  We have noway of stopping tesl at all so I believe we have a bigger fish to fry before handling WaitGroups.

**Comment by aghuna** [comment]
Added usage docs to README cf3ca385ece9294a49e4f623f4b77b38729f191d

**Pull request merged**